### DISCUSSION TITLE
Close button event

### DISCUSSION THREAD

**Discussion by amit7775** [unoplatform/uno#8520] [Q&A]
I want to write some piece of code on close button event. But i am unable to get close button event in WINUI 3 with UNO on visual studio 2022.

**Comment by jeromelaban** [comment]
Thanks for the question @amit7775, can you specify which platform you are targeting? Windows, Linux, Wasm, or other?

**Reply by amit7775** [reply]
Windows

**Reply by jeromelaban** [reply]
Thanks. Have you taken a look at [AppWindow](https://docs.microsoft.com/en-us/uwp/api/windows.ui.windowmanagement.appwindow?view=winrt-22000#events)?

**Reply by amit7775** [reply]
I had used App_CloseRequested button event on close(X) button but its not working. I had used this event in Mainpage.XAML.cs page.

**Reply by jeromelaban** [reply]
Thanks. Could you post a sample of what you tried?

**Reply by amit7775** [reply]
private async void App_CloseRequested(object sender, SystemNavigationCloseRequestedPreviewEventArgs e)
        {
}
I want to write some code on close button(X) click but event not fired on above close event. i had tried this  closerequested event in APP.xaml.cs and MainPage.xaml.cs both but on click of close button above event is not fired.

**Reply by jeromelaban** [reply]
Thanks, can you post a full sample of what you tried? Create a new Uno project from the template, adjust with what you tried, zip the project (without the bin/obj folders), and post it here.

**Reply by amit7775** [reply]
[UnoPOCApp.zip](https://github.com/unoplatform/uno/files/8469895/UnoPOCApp.zip)
PFA the sample code. button event code written in MainPage.XAML.CS page. The issue is App_CloseRequested event not fired on close(X) button.

**Reply by jeromelaban** [reply]
The `App_CloseRequested` is not going to be called as there nothing calling it.

You can use this:
```csharp
(Application.Current as App).MainWindow.Closed += (s, e) => { /* handle closed */ };
```
Or this:
```csharp
IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle((Application.Current as App).MainWindow);
 WindowId myWndId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd);
_appWindow = AppWindow.GetFromWindowId(myWndId);
_appWindow.Closing += (s, e) => { /* Handle closing */ };
```

In general though, for WinAppSDK on Windows questions, you should be asking directly in the Microsoft WinUI/WinAppSDK forums.

**Reply by amit7775** [reply]
Thanks,
(Application.Current as App).MainWindow.Closed += (s, e) => { /* handle closed */ };
This event works on close button when i put this code in MainPage.XAML.CS page in MainPage() constructor.
But icon residing in system tray for couple of second only. I think some other system in build close event suppress/override the above event.
Please let me know where we can use this event.
